src/passes/Monomorphize.cpp

```diff
@@ -47,44 +81,218 @@
-//       function references, if we monomorphized we'd end up specializing qsort
-//       for the particular functions it is given.
 //
 
 #include "ir/cost.h"
 #include "ir/find_all.h"
+#include "ir/manipulation.h"
 #include "ir/module-utils.h"
 #include "ir/names.h"
 #include "ir/type-updating.h"
 #include "ir/utils.h"
 #include "pass.h"
+#include "support/hash.h"
 #include "wasm-type.h"
 #include "wasm.h"
 
 namespace wasm {
 
 namespace {
 
+// Relevant information about a callsite for purposes of monomorphization.
+struct CallContext {
+  // The operands of the call, processed to leave the parts that make sense to
+  // keep in the context. That is, the operands of the CallContext are the exact
+  // code that will appear at the start of the monomorphized function. For
+  // example:
+  //
+  //  (call $foo
+  //    (i32.const 10)
+  //    (..something complicated..)
+  //  )
+  //  (func $foo (param $int i32) (param $complex f64)
+  //    ..
+  //
+  // The context operands are
+  //
+  //  [
+  //    (i32.const 10)       ;; Unchanged: this can be pulled into the called
+  //                         ;; function, and removed from the caller side.
+  //    (local.get $0)       ;; The complicated child cannot; keep it as a value
+  //                         ;; sent from the caller, which we will local.get.
+  //  ]
+  //
+  // Both the const and the local.get are simply used in the monomorphized
+  // function, like this:
+  //
+  //  (func $foo-monomorphized (param $0 ..)
+  //    (..local defs..)
+  //    ;; Apply the first operand, which was pulled into here.
+  //    (local.set $int
+  //      (i32.const 10)
+  //    )
+  //    ;; Read the second, which remains a parameter to the function.
+  //    (local.set $complex
+  //      (local.get $0)
+  //    )
+  //    ;; The original body.
+  //    ..
+  //
+  // The $int param is no longer a parameter, and it is set in a local at the
+  // top: we have "reverse-inlined" code from the calling function into the
+  // caller, pulling the constant 10 into here. The second parameter cannot be
+  // pulled in, so we must still send it, but we still have a local.set there to
+  // copy it into a local (this does not matter in this case, but does if the
+  // sent value is more refined; always using a local.set is simpler and more
+  // regular).
+  std::vector<Expression*> operands;
+
+  // Whether the call is dropped. TODO
+  bool dropped;
+
+  bool operator==(const CallContext& other) const {
+    // We consider logically equivalent expressions as equal (rather than raw
+    // pointers), so that contexts with functionally identical shape are
+    // treated the same.
+    if (operands.size() != other.operands.size()) {
+      return false;
+    }
+    for (Index i = 0; i < operands.size(); i++) {
+      if (!ExpressionAnalyzer::equal(operands[i], other.operands[i])) {
+        return false;
+      }
+    }
+
+    return dropped == other.dropped;
```

Comment:
Maybe check this before the operands because it is so cheap?

Reply:
Good idea, done.